xinference/model/llm/ggml/ctransformers.py

```diff
@@ -0,0 +1,292 @@
+
+    def _sanitize_model_config(
+        self, model_path, ctransformers_model_config: Optional[CtransformersModelConfig]
+    ) -> "AutoConfig":
+        try:
+            from ctransformers import AutoConfig, Config
+        except ImportError:
+            error_message = (
+                "Failed to import module 'ctransformers - AutoConfig and Config'"
+            )
+
+            installation_guide = [
+                f"Please make sure 'ctransformers' is installed.",
+                f"You can install it by checking out the repository for command:"
+                f"https://github.com/marella/ctransformers",
+            ]
+
+            raise ImportError(f"{error_message}\n\n{''.join(installation_guide)}")
+
+        # if the model have customized config, we update it.
+        if ctransformers_model_config:
+            potential_context_length = ctransformers_model_config.pop("n_ctx", None)
+            potential_gpu_layers = ctransformers_model_config.pop("n_gpu_layers", None)
+
+            if potential_context_length and potential_gpu_layers:
+                ctransformers_model_config_returned = Config(
+                    context_length=potential_context_length,
+                    gpu_layers=potential_gpu_layers,
+                )
+            elif potential_gpu_layers:
+                ctransformers_model_config_returned = Config(
+                    gpu_layers=potential_gpu_layers
+                )
+            elif potential_context_length:
+                ctransformers_model_config_returned = Config(
+                    context_length=potential_context_length
+                )
+            else:
+                ctransformers_model_config_returned = Config()
+        else:
+            ctransformers_model_config_returned = Config()
+
+        return AutoConfig(ctransformers_model_config_returned)
+
+    def _sanitize_generate_config(
+        self,
+        ctransformers_generate_config: Optional[CtransformersGenerateConfig],
+    ) -> CtransformersGenerateConfig:
+        try:
+            from ctransformers import Config
+        except ImportError:
+            error_message = "Failed to import module 'ctransformers - Config'"
+
+            installation_guide = [
+                f"Please make sure 'ctransformers' is installed.",
+                f"You can install it by checking out the repository for command:"
+                f"https://github.com/marella/ctransformers",
+            ]
+
+            raise ImportError(f"{error_message}\n\n{''.join(installation_guide)}")
+
+        # if the input config is not None, we try to copy the selected attributes to the ctransformersGenerateConfig.
+        if ctransformers_generate_config is None:
+            ctransformers_generate_config = CtransformersGenerateConfig()
+
+        # get the newest configuration from ctransformers.
+        default_config = Config()
+
+        ctransformers_generate_config.setdefault("top_k", default_config.top_k)
```

Comment:
No need to set defaults.

Reply:
all set default was removed except for threads, which set to 4